src/main/scala/viper/silver/ast/utility/QuantifiedPermissions.scala

```diff
@@ -244,6 +244,19 @@ object QuantifiedPermissions {
 
             desugarSourceQuantifiedPermissionSyntax(Forall(vars ++ nestedVars, combinedTriggers, Implies(newCond, nestedRhs)(rhs.pos, rhs.info, rhs.errT))(source.pos,MakeInfoPair(source.info, nested.info),MakeTrafoPair(source.errT,nested.errT)))
 
+          case lt@Let(v, e, bod) => {
+            val forallWithoutLet = Forall(vars, triggers, Implies(cond, bod)(rhs.pos, rhs.info))(source.pos, source.info)
+            // desugar the let-body
+            val desugaredWithoutLet = desugarSourceQuantifiedPermissionSyntax(forallWithoutLet)
+            desugaredWithoutLet.map{
+              case SourceQuantifiedPermissionAssertion(iqp, Implies(icond, irhs)) if (!irhs.isPure) =>
+                // Since the rhs cannot be a let-binding, we expand the let-expression
+                Forall(iqp.variables, iqp.triggers.map(t => t.replace(v.localVar, e)), Implies(icond, irhs.replace(v.localVar, e))(iqp.pos, iqp.info))(iqp.pos, iqp.info)

```

Comment:
This transformation fails on the following example:

```
predicate p(x: Ref) { acc(x.f) }

function pVal(x: Ref): Ref
  requires p(x)
{ unfolding p(x) in x.f }

field f: Ref

method bar(x: Seq[Ref]) 
  requires forall i: Int, j: Int :: 0 <= i && i < |x| && 0 <= j && j < |x| && i != j ==> x[i] != x[j]
  requires forall i: Int :: 0 <= i && i < |x| ==> p(x[i])
  requires forall i: Int, j: Int :: {x[i], x[j]} 0 <= i && i < |x| && 0 <= j && j < |x| && i != j ==> pVal(x[i]) != pVal(x[j])
  requires forall i: Int :: 0 <= i && i < |x| ==> let y == (pVal(x[i])) in p(unfolding p(x[i]) in y) // fails
```

Here the let binding cannot be replaced by a direct substitution, as this results in `forall i: Int :: 0 <= i && i < |x| ==> p(unfolding p(x[i]) in pVal(x[i])) `, which fails. It seems to me the problem are unfoldings in the arguments of accessibility predicates only. Unfoldings in the pure parts of the quantifier body get the forall binding reapplied. 

That said, the current transformation would already be a huge help for us. I'd be happy to merge this PR and to create a separate issue (or extend the existing issues) with my aforementioned case.